apps/faucet/src/endpoints/faucet.service.ts

```diff
@@ -0,0 +1,221 @@
+    if (!this.isFaucetEnabled()) {
+      throw new Error('Faucet not enabled');
+    }
+
+    if (!this.signer) {
+      throw new Error('No signer initialized');
+    }
+
+    if (!this.provider) {
+      throw new Error('No provider initialized');
+    }
+
+    const networkConfig = await this.getNetworkConfig();
+    if (!networkConfig) {
+      throw new Error('Network config could not be fetched');
+    }
+
+    if (!this.shouldBypassCaptchaCheck() && captcha) {
+      const success = await this.validateRecaptcha(captcha, clientIp);
+      if (!success) {
+        throw new BadRequestException('Failed captcha check');
+      }
+    }
+
+    if (address !== this.apiConfigService.config.faucetAddress) {
+      const isAddressUsed = await this.cachingService.get(CacheInfo.FaucetAddress(address).key);
+      if (isAddressUsed) {
+        return false;
+      }
+    }
+
+    const tx = new Transaction({
+      chainID: networkConfig.ChainID,
+      gasLimit: networkConfig.MinGasLimit,
+      gasPrice: networkConfig.MinGasPrice,
+      receiver: new Address(address),
+      value: this.apiConfigService.config.faucetAmount,
+      sender: new Address(this.apiConfigService.config.faucetAddress),
+    });
+
+    const txNonce = nonce ?? await this.getNonce();
+
+    this.logger.log(`Sending faucet xEGLD to address '${address}' with nonce '${txNonce}', ip '${clientIp}', chain ID '${networkConfig.ChainID.valueOf()}', captcha '${captcha}'`);
+
+    tx.setNonce(txNonce);
+
+    const signature = await this.signer.sign(tx.serializeForSigning());
+    tx.applySignature(signature);
+
+    await this.provider.sendTransaction(tx);
+
+    const faucetToken = this.apiConfigService.config.faucetToken;
+    if (faucetToken && nonce === undefined) {
+      const tx2 = this.getTransaction(address, faucetToken, networkConfig);
+
+      const nonce2 = await this.getNonce();
+      tx2.setNonce(nonce2);
+
+      this.logger.log(`Sending faucet token to address '${address}' with nonce '${nonce2}', ip '${clientIp}'`);
+
+      const signature2 = await this.signer.sign(tx2.serializeForSigning());
+      tx2.applySignature(signature2);
+
+      await this.provider.sendTransaction(tx2);
+    }
+
+    await this.cachingService.set(
+      CacheInfo.FaucetAddress(address).key,
+      true,
+      Constants.oneSecond() * this.apiConfigService.config.faucetCooldownSameAddressInSec,
+      );
+
+    return true;
+  }
+
+  private getTransaction(address: string, faucetToken: string, networkConfig: NetworkConfig): Transaction {
+    const tokenHex = this.padHex(Buffer.from(faucetToken.split('-').slice(0, 2).join('-'), 'ascii').toString('hex'));
+    const tokenAmountHex = this.padHex(BigInt(this.apiConfigService.config.faucetTokenAmount).toString(16));
+
+    const isNft = faucetToken.split('-').length === 3;
+    if (isNft) {
+      const tokenNonceHex = faucetToken.split('-')[2];
+      return new Transaction({
```

Comment:
the logic for constructing the Transaction object is almost the same for both cases (NFT and token ). maybe you can extract the common properties to reduce duplication and make the code more maintainable .

Reply:
done. extracted different fields (receiver and data) into variables and I only construct once the transaction